resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterSchedulerSource.scala

```diff
@@ -34,91 +36,103 @@ private[mesos] class MesosClusterSchedulerSource(scheduler: MesosClusterSchedule
-  // Duration from initial submission to finished, regardless of retries
-  private val submitToFinish = metricRegistry.timer(MetricRegistry.name("submitToFinish"))
-  // Duration from initial submission to failed, regardless of retries
-  private val submitToFail = metricRegistry.timer(MetricRegistry.name("submitToFail"))
-
-  // Duration from (most recent) launch to finished
-  private val launchToFinish = metricRegistry.timer(MetricRegistry.name("launchToFinish"))
-  // Duration from (most recent) launch to retry
-  private val launchToRetry = metricRegistry.timer(MetricRegistry.name("launchToRetry"))
+  private val submitToFirstLaunch =
+    metricRegistry.timer(MetricRegistry.name("driver", "submit_to_first_launch"))
+  private val submitToException =
+    metricRegistry.timer(MetricRegistry.name("driver", "submit_to_exception"))
+  private val launchToRetry = metricRegistry.timer(MetricRegistry.name("driver", "launch_to_retry"))
 
   // Histogram of retry counts at retry scheduling
-  private val retryCount = metricRegistry.histogram(MetricRegistry.name("retryCount"))
+  private val retryCount = metricRegistry.histogram(MetricRegistry.name("driver", "retry_counts"))
 
   // Records when a submission initially enters the launch queue.
   def recordQueuedDriver(): Unit = queuedCounter.inc()
 
-  // Records when a submission has failed an attempt and is marked for retrying.
+  // Records when a submission has failed an attempt and is eligible to be retried
   def recordRetryingDriver(state: MesosClusterSubmissionState): Unit = {
     state.driverDescription.retryState.foreach(retryState => retryCount.update(retryState.retries))
-    recordTimeSince(launchToRetry, state.startDate)
+    // Duration from (most recent) launch to retry
+    recordTimeSince(state.startDate, launchToRetry)
     retryingCounter.inc()
   }
 
   // Records when a submission is launched.
   def recordLaunchedDriver(desc: MesosDriverDescription): Unit = {
     if (!desc.retryState.isDefined) {
-      recordTimeSince(submitToFirstLaunch, desc.submissionDate)
+      // Duration from submission to FIRST/NON-RETRY launch
+      recordTimeSince(desc.submissionDate, submitToFirstLaunch)
     }
     launchedCounter.inc()
   }
 
-  // Records when a submission has successfully finished.
-  def recordFinishedDriver(state: MesosClusterSubmissionState): Unit = {
-    recordTimeSince(submitToFinish, state.driverDescription.submissionDate)
-    recordTimeSince(launchToFinish, state.startDate)
-    finishedCounter.inc()
+  // Records when a submission has successfully finished, or failed and was not eligible for retry.
+  def recordFinishedDriver(state: MesosClusterSubmissionState, mesosState: MesosTaskState): Unit = {
+    // Record against task state, e.g. "finished_driver_count.task_lost"
+    val mesosStateStr = mesosState.name().toLowerCase()
+
+    // Duration from initial submission to finished, regardless of retries
```

Comment:
TODO: These metrics don't appear to show up. Guessing that the metricRegistry calls after initialization are resulting in these values getting dropped on the floor.

Reply:
Yes, that's what's happening. There's a root-level `MetricRegistry` in spark's `MetricsSystem.scala`. Everything needs to be pre-initialized before the `MesosClusterSchedulerSource` is passed there.